### PR TITLE
Allow additional Config Options Part 1 : Config, ConfigLoader, ConfigFormat and ConfigFactory

### PR DESCRIPTION
**Pull request by chesterxgchen**
### Description
**Goals:**
In addition to the current JSON configuration for jobs, system are able to choose alternative configurations formats such as Pyhocon (python HOCON, considered it as superset of json) and Omegaconf ( considered it as superset of YAML)

This will allow user to add

Comment support ( inline, block)
nested configuration structure
ENV variable substitution
variable substitution
Multi-line strings ( for pyhocon)
optional overwrite from env. Variables (pyhocon)
Dictionary, list merge
Include directive support (pyhocon)
Default support API
dot notation for access nested elements

Non-goals:
Replace current configuration.

Design Changes:
1. new existing configuration or even constant changes. 
2. For file-configuration, we are going to search based on the order of ".json", ".conf", ".yml" extensions. The first found config file base name such <filename>.ext will be used as the configuration. 
3. We can mixed different configuration formats as one wishes. 
4. This PR introduce the Config (with JsonConfig, PyhoconConfig, OmegaConf), ConfigFormat, ConfigLoader, ConfigFactory. 
5. OmegaConf is not yet implemented, right now is just placeholder. 

for detailed discussion and design, please look at [design doc](https://docs.google.com/document/d/1t_1BjgQ6OCxw2LpaIReuXhtvIBnh7NDU7a0amGY23c8)

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: nvflare/fuel/utils/config.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    @classmethod
+    def config_exts(cls) -> str:
+        search_sequence: List[ConfigFormat] = cls.search_order()
+        extensions = []
+        for fmt in search_sequence:
+            exts = cls.get_extensions(fmt.name)
+            if exts:
+                extensions.extend(exts)
+
+        return "|".join(iter(extensions))
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/config.py:62]: why do we need this method?

File: nvflare/fuel/utils/config.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    @abstractmethod
+    def get_format(self):
+        pass
+
+    @abstractmethod
+    def get_location(self) -> Optional[str]:
+        # return the real file path of the reel configuration.
+        # For example, if initial config_fed_json.json was initially input file path. but the real file path is
+        # config_fed_json.conf. the get location will be return config_fed_json.conf
+        pass
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/config.py:83]: Can we improve the docstring?

This method is not clear to me, as the implementation always return "self.file_path"

Comment (YuanTingHsieh) [reply nvflare/fuel/utils/config.py:83]: If get_location return None, what does that mean?

File: nvflare/fuel/utils/config.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    @abstractmethod
+    def to_conf(self, element: Dict) -> str:
+        pass
+
+    @abstractmethod
+    def get_int(self, key: str, default=None) -> Optional[int]:
+        pass
+
+    @abstractmethod
+    def get_float(self, key: str, conf=None, default=None) -> Optional[float]:
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/config.py:98]: should the signature be consistent with other "get" method?

File: nvflare/fuel/utils/config.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    @abstractmethod
+    def get_conf(self):
+        # Return the original underline config object representation if you prefer to use it directly
+        # Pyhocon → ConfigTree
+        # JSON → Dict
+        # OMEGACONF → Conf
+        pass
+
+    @abstractmethod
+    def get_format(self):
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/config.py:75]: should this return ConfigFormat?

File: nvflare/fuel/utils/config.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    def get_str(self, key: str, default=None) -> Optional[str]:
+        pass
+
+    @abstractmethod
+    def get_list(self, key: str, default=None) -> Optional[List]:
+        pass
+
+    @abstractmethod
+    def get_config(self, key: str, default=None):
+        pass
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/config.py:115]: What is the difference "get_config" VS "get_conf" VS "get_location"?


Say we have a "config_fed_server.json"

get_config will return "config_fed_server.json"?
get_conf will return a dictionary
get_location will return "config_fed_server.json"?

so get_config and get_location is the same thing?

File: nvflare/fuel/utils/config_factory.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    @staticmethod
+    def search_config_format(
+        init_file_path, search_dirs: Optional[List[str]] = None
+    ) -> (Optional[ConfigFormat], Optional[str]):
+        # we ignore the original extension
+        logger = ConfigFactory.logger
+        if not search_dirs:
+            parent_dir = pathlib.Path(init_file_path).parent
+            search_dirs = [str(parent_dir)]
+        file_path = os.path.splitext(pathlib.Path(init_file_path).name)[0]
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/config_factory.py:38]: the comment "# we ignore the original extension" should be above this line ^

Comment (YuanTingHsieh) [reply nvflare/fuel/utils/config_factory.py:38]: suggest renaming file_path to maybe file_name as it is not path anymore.

File: nvflare/fuel/utils/config.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    def ordered_search_extensions(cls) -> List:
+        search_sequence: List[ConfigFormat] = cls.search_order()
+        extensions = []
+        for fmt in search_sequence:
+            exts = cls.get_extensions(fmt.name)
+            if exts:
+                for ext in exts:
+                    extensions.append((fmt, ext))
+
+        return extensions
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/config.py:47]: Why do we put these methods in ConfigFormat but not ConfigFactory?

File: nvflare/fuel/utils/json_config_loader.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        return json.dumps(element)
+
+    def get_int(self, key: str, default=None) -> Optional[int]:
+        value = self.conf.get(key, default)
+        try:
+            return int(value) if value is not None else None
+        except (TypeError, ValueError):
+            raise ConfigError("{key} has type '{type}' rather than 'int'".format(key=key, type=type(value).__name__))
+
+    def get_float(self, key: str, conf=None, default=None) -> Optional[float]:
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/json_config_loader.py:52]: method signature should be consistent with other get

File: nvflare/fuel/utils/json_config_loader.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        value = self.conf.get(key, default)
+        if value:
+            if isinstance(value, Dict):
+                return JsonConfig(value, self.file_path)
+            else:
+                raise ConfigError(
+                    "{key} has type '{type}' rather than 'Dict' or 'JsonConfig' ".format(
+                        key=key, type=type(value).__name__
+                    )
+                )
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/json_config_loader.py:104]: if value is Dict, things are good,
if value is of type "JsonConfig", this code will raise ConfigError

File: nvflare/fuel/utils/omegaconf_loader.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Dict, Optional
+
+from nvflare.fuel.utils.config import Config, ConfigFormat, ConfigLoader
+
+
+class OmegaConfLoader(ConfigLoader):
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/omegaconf_loader.py:19]: maybe we add this file in another follow up PR.

File: nvflare/fuel/utils/pyhocon_loader.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        return self._convert_conf_item(self.conf)
+
+    def to_conf(self, element: Dict) -> str:
+        config = CF.from_dict(element)
+        return HOCONConverter.to_hocon(config)
+
+    def get_int(self, key: str, default=None) -> Optional[int]:
+        return self.conf.get_int(key, default)
+
+    def get_float(self, key: str, conf=None, default=None) -> Optional[float]:
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/pyhocon_loader.py:50]: suggest change it to be consistent with other get

File: nvflare/fuel/utils/config.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+    @classmethod
+    def config_exts(cls) -> str:
+        search_sequence: List[ConfigFormat] = cls.search_order()
+        extensions = []
+        for fmt in search_sequence:
+            exts = cls.get_extensions(fmt.name)
+            if exts:
+                extensions.extend(exts)
+
+        return "|".join(iter(extensions))
```
Comment (yanchengnv) [inline nvflare/fuel/utils/config.py:50]: Why joining into a str, instead of returning the list?

File: nvflare/fuel/utils/config.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+            element: dict
+
+        Returns: str string representation of the configuration in given format
+
+        """
+
+        pass
+
+    @abstractmethod
+    def get_int(self, key: str, default=None) -> Optional[int]:
```
Comment (yanchengnv) [inline nvflare/fuel/utils/config.py:109]: Why do we need these get_xxx() methods?

I thought the conf loader will only return a dict. It will be up to another layer (i.e. ConfigService) to do all these get_xxx() stuff.

File: nvflare/fuel/utils/config.py [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+        pass
+
+    @abstractmethod
+    def get_list(self, key: str, default=None) -> Optional[List]:
+        pass
+
+    @abstractmethod
+    def get_config(self, key: str, default=None):
+        """
+           This method returns nvflare.fuel.utils.Config object
```
Comment (yanchengnv) [inline nvflare/fuel/utils/config.py:131]: Not sure I get this. What is nvflare.fuel.utils.Config object? Isn't it the same as the "self"?

File: nvflare/fuel/utils/config_factory.py [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+        """
+        logger = ConfigFactory.logger
+        if not search_dirs:
+            parent_dir = pathlib.Path(init_file_path).parent
+            search_dirs = [str(parent_dir)]
+        # we ignore the original extension
+        file_basename = os.path.splitext(pathlib.Path(init_file_path).name)[0]
+        for search_dir in search_dirs:
+            logger.debug(f"search file:{file_basename} basename, search dirs = {search_dirs}")
+            for root, dirs, files in os.walk(search_dir):
```
Comment (yanchengnv) [inline nvflare/fuel/utils/config_factory.py:66]: Should not walk thru all sub folders, otherwise a wrong file could be found. Instead, only look at the files directly in the dir.
Also the extension sequence is not right - see my comment before about non-default should be searched first before searching default cross ext types.

File: nvflare/fuel/utils/config_factory.py [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+
+    @staticmethod
+    def has_config(init_file_path: str, search_dirs: Optional[List[str]] = None) -> bool:
+        _, real_file_path = ConfigFactory.search_config_format(init_file_path, search_dirs)
+        return real_file_path is not None
+
+    @staticmethod
+    def match_config(parent, init_file_path, match_fn) -> bool:
+        # we ignore the original extension
+        basename = os.path.splitext(init_file_path)[0]
```
Comment (yanchengnv) [inline nvflare/fuel/utils/config_factory.py:91]: What is basename has "." in it?

**Review by yanchengnv — CHANGES_REQUESTED** [review]
Looks good overall. Some changes needed:
- The search order of non-default vs. default
- I don't think get_xxx() methods should be part of the Config. Var handling should be handled by another layer. This PR should simply support loading of different formats and return dict. It's up to another layer how to determine the value of a var based on the config and other factors.

File: nvflare/fuel/utils/config.py (hunk 12 above)
Comment (yanchengnv) [reply nvflare/fuel/utils/config.py:50]: You could also directly use the list in your error message. It will print quite nicely as a list.

File: nvflare/fuel/utils/config.py (hunk 14 above)
Comment (yanchengnv) [reply nvflare/fuel/utils/config.py:131]: But I am not sure this is needed. If every loader simply returns a dict, the nested handling is only on dict. There is a dict_utils already. We could add more funcs there if needed. We don't need every Config type to do it.

File: nvflare/fuel/utils/config_factory.py (hunk 16 above)
Comment (yanchengnv) [reply nvflare/fuel/utils/config_factory.py:91]: Right, splitext doesn't know ".json.default" should be treated as one ext.
I wonder maybe we should change the treatment of default files. Instead of xxx.json.default, why don't we change it to default.xxx.json? This will make it much easier for you, right?
This is totally under our control. Of course we will need to change the Provision to produce default files with the new names.

Comment (yanchengnv) [reply nvflare/fuel/utils/config_factory.py:91]: Also xxx.json.default is anti-pattern - technically the ext here is ".default". Many tools rely on ext to work.

File: nvflare/fuel/utils/config.py (hunk 13 above)
Comment (YuanTingHsieh) [reply nvflare/fuel/utils/config.py:109]: ^ "the get_xxx just enforce the types or interpret the types."

So you mean for the same key say "num_of_rounds", we could use config.get_int("num_of_rounds") to get an integer representation and we could also use config.get_float("num_of_rounds") to get as float?

Then the ConfigService's get_xxx will try to 
1. get this key from environment variable
2. get this key from config_fed_server.json using "config.get_xxx"
3. get this key from resources.json using "config.get_xxx"
4. get this key from command line arguments

^ the order can be what we decide.
The first value return will be the value that is used.

Is this what we are thinking?

Comment (yanchengnv) [reply nvflare/fuel/utils/config.py:109]: Yes and no. Even though each config could provide its own get_xxx(), it's unnecessary since everything is loaded into dict already, and all vars could be retrieved from the dict, regardless where the dict is loaded from.

Secondly, as to how the var's value is determined, it could vary, depending on the nature of the var. What you laid out here is just one possible path. There are other cases.

Bottom line is that the var value determination logic is beyond a specific config loader.

File: nvflare/fuel/utils/config.py [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+        Args:
+            config_str:
+        Returns:nvflare.fuel.utils.Config
+
+        """
+        raise NotImplementedError
+
+    def load_config_from_dict(self, config_dict: dict) -> Config:
+        """
+        Load Configuration based for given config dict.
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/config.py:127]: Thanks for the update, the docstrings and API layer is clearer to me right now.

For the docstrings, can you follow the google docstring style here: https://google.github.io/styleguide/pyguide.html#383-functions-and-methods 

Thanks!

File: nvflare/fuel/utils/config.py [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+        convert underline config object to dictionary
+        Returns:Dict
+
+        """
+        pass
+
+    @abstractmethod
+    def to_conf(self, element: Dict) -> str:
+        """
+            convert dict element to the str representation of the underline configuration.
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/config.py:82]: Since it returns a string, suggest change method name to "to_conf_str" or "to_str" to be more clear?

File: tests/unit_test/fuel/utils/json_config_test.py [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+        assert config is not None
+        conf = config.get_native_conf()
+        assert conf["a"]["a1"] == 200
+        assert conf.get("b") == 1
+        assert conf.get("c") == "hello"
+        assert conf.get("d") == [200, 400, 500]
+        assert conf.get("e1") == "Yes"
+        assert conf.get("e2") == "True"
+        assert conf.get("e3") == "NO"
+        assert conf.get("e4", None) is None
```
Comment (YuanTingHsieh) [inline tests/unit_test/fuel/utils/json_config_test.py:68]: Would "assert conf == dicts" achieve the same effect?

File: nvflare/fuel/utils/config.py [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+        configuration from overwrite_config if provided will be the merge with config overwrite the same key
+        Args:
+            file_path: file path for configuration to be loaded
+            default_file_path: file path for default configuration
+            overwrite_config: dict config that will  overwrite the final config if provided
+
+        Returns:nvflare.fuel.utils.Config
+        """
+
+        pass
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/config.py:112]: Another approach:
What if we have "ConfigLoader" just load the config, so `def load_config(self, file_path)`.
And we let `ConfigFactory` to determine what config should overwrite the other?

File: tests/unit_test/fuel/utils/json_config_test.py (hunk 19 above)
Comment (YuanTingHsieh) [reply tests/unit_test/fuel/utils/json_config_test.py:68]: I re-read again and found your loader.load_config not only loads config from file it also loads from dict and it also merges them.
Given that implementation this test case is good.

File: nvflare/fuel/utils/config.py [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+            config_str:
+        Returns:
+            return Config
+
+        """
+        raise NotImplementedError
+
+    def load_config_from_dict(self, config_dict: dict) -> Config:
+        """
+            Load Configuration based for given config dict.
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/config.py:141]: Instead of starting in the next line, it should starts right after """.

For example:  `"""Loads Configuration based for given config dict.`

File: nvflare/fuel/utils/config.py [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+
+        Returns:
+            return Config
+        """
+
+        pass
+
+    def load_config_from_str(self, config_str: str) -> Config:
+        """
+            Load Configuration based on the string representation of the underline configuration
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/config.py:128]: Same here for the style.

File: nvflare/fuel/utils/config.py [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+                ".yml.default": ConfigFormat.OMEGACONF,
+            }
+        )
+
+
+class Config(ABC):
+    @abstractmethod
+    def get_native_conf(self):
+        """
+            Return the original underline config object representation if you prefer to use it directly
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/config.py:44]: docstring style

**Review by YuanTingHsieh — COMMENTED** [review]
A few places docstring style needs to be corrected.

Otherwise the logic LGTM.

**Review by YuanTingHsieh — APPROVED** [review]
I still feel the "load_config" method of "ConfigLoader" should be keep simple.
I think it should not also do merge or replacing.

But we can discuss that further in part2 when I see more examples.

**Review by yanchengnv — CHANGES_REQUESTED** [review]
The main issue is where the merge logic belongs. I don't think loaders should worry about it.

**Review by yanchengnv — APPROVED** [review]
LGTM
